File: src/shared/java/Logging.java
package com.tenstorrent.shared;

import java.util.Locale;
import java.util.logging.ConsoleHandler;
import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public final class Logging {
    private static final String LEVEL_ENV = "TT_LOG_LEVEL";

    private Logging() {}

    public static Level resolveLevel(String overrideLevel) {
        String candidate = (overrideLevel == null || overrideLevel.isBlank())
                ? System.getenv(LEVEL_ENV)
                : overrideLevel;

        if (candidate == null || candidate.isBlank()) {
            return Level.INFO;
        }

        String normalized = candidate.trim().toUpperCase(Locale.ROOT);
        return switch (normalized) {
            case "TRACE", "FINEST" -> Level.FINEST;
            case "DEBUG", "FINE" -> Level.FINE;
            case "INFO", "INFORMATION" -> Level.INFO;
            case "WARN", "WARNING" -> Level.WARNING;
            case "ERROR", "SEVERE" -> Level.SEVERE;
            default -> Level.INFO;
        };
    }

    public static Logger getLogger(String name) {
        Logger logger = Logger.getLogger(name);
        Level level = resolveLevel(null);
        logger.setLevel(level);
        logger.setUseParentHandlers(false);

        boolean hasHandlers = logger.getHandlers().length > 0;
        if (!hasHandlers) {
            ConsoleHandler handler = new ConsoleHandler();
            handler.setLevel(level);
            handler.setFormatter(new SimpleFormatter());
            logger.addHandler(handler);
        } else {
            for (Handler handler : logger.getHandlers()) {
                handler.setLevel(level);
            }
        }

        return logger;
    }
}

File: src/shared/dotnet/Logging.cs
using System;
using Microsoft.Extensions.Logging;

namespace Tenstorrent.Shared
{
    public static class Logging
    {
        private const string LevelEnv = "TT_LOG_LEVEL";

        public static LogLevel ResolveLogLevel(string? overrideLevel = null)
        {
            var candidate = string.IsNullOrWhiteSpace(overrideLevel)
                ? Environment.GetEnvironmentVariable(LevelEnv)
                : overrideLevel;

            if (string.IsNullOrWhiteSpace(candidate))
            {
                return LogLevel.Information;
            }

            return candidate.Trim().ToUpperInvariant() switch
            {
                "TRACE" => LogLevel.Trace,
                "DEBUG" => LogLevel.Debug,
                "INFO" or "INFORMATION" => LogLevel.Information,
                "WARN" or "WARNING" => LogLevel.Warning,
                "ERROR" => LogLevel.Error,
                "CRITICAL" or "FATAL" => LogLevel.Critical,
                _ => LogLevel.Information,
            };
        }

        public static ILoggerFactory CreateLoggerFactory(LogLevel? minimumLevel = null)
        {
            var resolved = minimumLevel ?? ResolveLogLevel();

            return LoggerFactory.Create(builder =>
            {
                builder.SetMinimumLevel(resolved);
                builder.AddSimpleConsole(options =>
                {
                    options.SingleLine = true;
                    options.TimestampFormat = "yyyy-MM-dd HH:mm:ss ";
                });
            });
        }
    }
}
